test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java

```diff
@@ -176,83 +179,177 @@ public TestCase nodes(int nodeCount)
-            this.upgrade.addAll(upgrade);
+            this.minimalApplicableVersion = minimalSupportedVersion;
+            this.maximalApplicableVersion = maximalSupportedVersion;
+            addUpgrades();
             return this;
         }
 
         /**
-         * performs all supported upgrade paths from the "from" target; example
-         * {@code upgradesFrom(4.0, 4.2); // produces: 4.0 -> 4.1, 4.0 -> 4.2}
-         */
-        public TestCase upgradesFrom(Semver from, Semver to)
+         * this is a deprecated, backwards-compatible API for upgrade tests, that allows to specify which upgrades to perform
+         *
+         * upgradesToCurrentFrom(Semver from) sets the upgrades to perform to be all supported direct upgrades (upgrade
+         * paths with length 1) to the CURRENT version, but omitting versions older than the given "from" version
+         *
+         * This intent is best expressed by the newer alternatives:
+         * - appplicableVersionsRange(from, CURRENT)
+         * or, perhaps, depending on the actual case:
+         * - minimalApplicableVersion(from)
+         *
+         * please note, that upgrades that do not involve the CURRENT version are not meant to be run in upgrade tests
+         *
+         * Beware, there may be no supported direct upgrade path from -> CURRENT, and this is OK.
+         * e.g. when CURRENT == v4
+         * {@code upgradesToCurrentStartingNoEarlierThan(3.0); // produces: 3.0 -> CURRENT, 3.11 -> CURRENT, …}
+         * and when CURRENT == v5
+         * {@code upgradesToCurrentStartingNoEarlierThan(3.0); // produces: 4.0 -> CURRENT, 4.1 -> CURRENT, …}
+         * and when CURRENT == v3.11
+         * {@code upgradesToCurrentStartingNoEarlierThan(3.0); // produces: 3.0 -> CURRENT, …}
+         **/
+        @Deprecated
+        public TestCase upgradesToCurrentFrom(Semver from)
         {
-            List<TestVersions> upgrade = new ArrayList<>();
-            NavigableSet<Semver> vertices = sortedVertices(SUPPORTED_UPGRADE_PATHS);
-            for (Semver end : vertices.subSet(from, false, to, true))
-            {
-                // only include pairs that are allowed, and start or end on CURRENT
-                if (SUPPORTED_UPGRADE_PATHS.hasEdge(from, end) && contains(from, end, CURRENT))
-                    upgrade.add(new TestVersions(versions.getLatest(from), Collections.singletonList(versions.getLatest(end))));
-            }
-            logger.info("Adding upgrades of\n{}", upgrade.stream().map(TestVersions::toString).collect(Collectors.joining("\n")));
-            this.upgrade.addAll(upgrade);
-            return this;
+            return applicableVersionsRange(from, CURRENT);
         }
 
         /**
-         * performs all supported upgrade paths that exist in between from and to that include the current version.
-         * This call is equivalent to calling {@code upgradesTo(from, CURRENT).upgradesFrom(CURRENT, to)}.
```

Comment:
it is very confusing that his javadoc says that this method is equivalent to sth, and then the method's code is different. There are no performance considerations etc.